frontend/src/app/workflows/workflows.component.ts

```diff
@@ -112,19 +109,20 @@ export class WorkflowsComponent implements OnInit {
     );
   }
 
-  setIndexFromIframe(index: number, diagramType: 'pages' | 'objects' | 'class') {
+  @HostListener('window:message', ['$event'])
+  handleGlobalMessages(event: MessageEvent) {
+    const messageData = JSON.parse(event.data);
     // It needs to be run in the NgZone because only then angular change detection gets a grip on the change
     this.zone.run(() => {
-      this.newPageIndex = index;
-      this.currentDisplay = diagramType;
+      this.newPageIndex = messageData.index;
+      this.currentDisplay = messageData.diagramType;
     });
-  }
 
-  changeFrameWithToast(toastContent: string, index: number) {
-    this.setIndexFromIframe(index, 'pages');
-    this.zone.run(() => {
-      this.toastService.success('Page Action', toastContent);
-    });
+    if (messageData.type === 'changeFrameWithToast') {
+      this.zone.run(() => {
```

Comment:
`@HostListener` already ensures this runs in the right zone.

Reply:
goodness gracious. thanks